chunk/chunk.go

```diff
@@ -23,11 +40,82 @@ func (cs ByID) Len() int           { return len(cs) }
 func (cs ByID) Swap(i, j int)      { cs[i], cs[j] = cs[j], cs[i] }
 func (cs ByID) Less(i, j int) bool { return cs[i].ID < cs[j].ID }
 
+// NewChunk creates a new chunk
+func NewChunk(fp model.Fingerprint, metric model.Metric, c *prom_chunk.Desc) Chunk {
+	return Chunk{
+		ID:       fmt.Sprintf("%d:%d:%d", fp, c.ChunkFirstTime, c.ChunkLastTime),
+		From:     c.ChunkFirstTime,
+		Through:  c.ChunkLastTime,
+		Metric:   metric,
+		Encoding: c.C.Encoding(),
+		Data:     c.C,
+	}
+}
+
+func (c *Chunk) reader() (io.ReadSeeker, error) {
+	// Encode chunk metadata into snappy-compressed buffer
+	var metadata bytes.Buffer
+	if err := json.NewEncoder(snappy.NewWriter(&metadata)).Encode(c); err != nil {
+		return nil, err
+	}
+
+	lenBytes := [4]byte{}
+	binary.BigEndian.PutUint32(lenBytes[:], uint32(metadata.Len()))
+
+	// TODO consider adding a .Reader() to upstream to remove copy
+	data := make([]byte, prom_chunk.ChunkLen)
+	if err := c.Data.MarshalToBuf(data); err != nil {
+		return nil, err
+	}
+
+	// Body is chunk bytes (uncompressed) with metadata appended on the end.
+	return ioutils.MultiReadSeeker(
+		bytes.NewReader(lenBytes[:]),
+		bytes.NewReader(metadata.Bytes()),
+		bytes.NewReader(data),
+	), nil
+}
+
+func (c *Chunk) decode(r io.Reader) error {
+	// Legacy chunks were written with metadata in the index.
+	if c.metadataInIndex {
+		var err error
+		c.Data, err = prom_chunk.NewForEncoding(prom_chunk.DoubleDelta)
+		if err != nil {
+			return err
+		}
+		return c.Data.Unmarshal(r)
+	}
+
+	var metadataLen uint32
+	if err := binary.Read(r, binary.BigEndian, &metadataLen); err != nil {
+		return err
+	}
+
+	err := json.NewDecoder(snappy.NewReader(&io.LimitedReader{
+		N: int64(metadataLen),
+		R: r,
+	})).Decode(c)
+	if err != nil {
+		return err
+	}
+
+	if c.Encoding == prom_chunk.Delta {
```

Comment:
I think the comment you made in the struct (about never using Delta) should be moved down here.